File: Sources/DiskMapCore/DuplicateFinder.swift
import Foundation
import CryptoKit

public struct DuplicateGroup: Sendable, Equatable {
    public let hash: String
    public let fileIDs: [Int32]
    public let sizeEach: Int64
    /// True when every file shares one physical extent map. Deleting one
    /// copy does not free `sizeEach` — the blocks stay until the last
    /// copy in this group is gone.
    public let sharesStorage: Bool

    public init(hash: String, fileIDs: [Int32], sizeEach: Int64, sharesStorage: Bool) {
        self.hash = hash
        self.fileIDs = fileIDs
        self.sizeEach = sizeEach
        self.sharesStorage = sharesStorage
    }

    /// Bytes a later confirm would actually free. Content copies each
    /// occupy their own blocks. Shared extents count once, and only when
    /// every copy in the group is being deleted.
    public func reclaimableBytes(deleting selected: Set<Int32>) -> Int64 {
        let removing = fileIDs.filter { selected.contains($0) }
        guard !removing.isEmpty, sizeEach > 0 else { return 0 }
        if sharesStorage {
            return removing.count == fileIDs.count ? sizeEach : 0
        }
        return sizeEach * Int64(removing.count)
    }

    /// Oldest modified day, then lowest id. That file stays unchecked so
    /// a one-click stage keeps one original.
    public func defaultKeeperID(modifiedDay: (Int32) -> Int32) -> Int32? {
        fileIDs.min { lhs, rhs in
            let leftDay = modifiedDay(lhs)
            let rightDay = modifiedDay(rhs)
            if leftDay != rightDay { return leftDay < rightDay }
            return lhs < rhs
        }
    }
}


public enum DuplicateScanPhase: String, Sendable, Equatable {
    case idle
    case preparing
    case collecting
    case grouping
    case hashing
    case assembling
    case complete
    case noResults
    case cancelled
    case failed

    public var title: String {
        switch self {
        case .idle: return "Ready"
        case .preparing: return "Preparing duplicate search…"
        case .collecting: return "Collecting candidate files…"
        case .grouping: return "Grouping by size…"
        case .hashing: return "Hashing colliding files…"
        case .assembling: return "Assembling duplicate groups…"
        case .complete: return "Finished"
        case .noResults: return "No duplicates found"
        case .cancelled: return "Cancelled"
        case .failed: return "Failed"
        }
    }
}

public struct DuplicateScanResult: Sendable, Equatable {
    public var groups: [DuplicateGroup]
    public var fullContentHashCalls: Int
}

/// Three-phase duplicate detection, cheapest checks first:
///
///   1. Group by exact logical size.
///   2. Within a size group, hash only the first 64 KB.
///   3. Only for files that still collide, hash full content —
///      unless `CloneDetector.areLikelyClones` says they share a full
///      extent map. That check is the packed `F_LOG2PHYS_EXT` map from
///      TASK-005, not a first-block offset. A clone overwritten past the
///      64 KB window still collides on the partial hash and must be
///      hashed; a first-extent match would have grouped it as identical.
public enum DuplicateFinder {

    public static func findDuplicates(
        candidates: [(id: Int32, url: URL, size: Int64)],
        progress: (@Sendable (DuplicateScanPhase, Int, Int) -> Void)? = nil
    ) async throws -> [DuplicateGroup] {
        try await scan(candidates, progress: progress).groups
    }

    /// Regular files under `root`, skipping directories, empty files, and
    /// not-downloaded iCloud placeholders (opening those would download).
    public static func candidates(in tree: FileTree, root: URL) -> [(id: Int32, url: URL, size: Int64)] {
        (try? cancellableCandidates(in: tree, root: root, progress: nil)) ?? []
    }

    /// Builds candidate paths away from the main actor and cooperates with
    /// cancellation during very large tree walks.
    public static func candidatesAsync(
        in tree: FileTree,
        root: URL,
        progress: (@Sendable (_ examined: Int, _ candidates: Int) -> Void)? = nil
    ) async throws -> [(id: Int32, url: URL, size: Int64)] {
        let worker = Task.detached(priority: .userInitiated) {
            try cancellableCandidates(in: tree, root: root, progress: progress)
        }
        return try await withTaskCancellationHandler {
            try await worker.value
        } onCancel: {
            worker.cancel()
        }
    }

    private static func cancellableCandidates(
        in tree: FileTree,
        root: URL,
        progress: (@Sendable (_ examined: Int, _ candidates: Int) -> Void)?
    ) throws -> [(id: Int32, url: URL, size: Int64)] {
        var result: [(id: Int32, url: URL, size: Int64)] = []
        guard tree.count > 0 else { return result }
        var stack: [Int32] = [0]
        var examined = 0
        while let id = stack.popLast() {
            if examined & 2_047 == 0 {
                try Task.checkCancellation()
                progress?(examined, result.count)
            }
            let index = Int(id)
            if index > 0 {
                let notDownloaded = tree.flags[index] & NodeFlags.notDownloaded != 0
                if !tree.isDirectory[index], !notDownloaded, tree.logicalSize[index] > 0 {
                    result.append((id, tree.path(of: id, root: root), tree.logicalSize[index]))
                }
            }
            var children: [Int32] = []
            var child = tree.firstChild[index]
            while child != -1 {
                children.append(child)
                child = tree.nextSibling[Int(child)]
            }
            stack.append(contentsOf: children.reversed())
            examined += 1
        }
        try Task.checkCancellation()
        progress?(examined, result.count)
        return result
    }

    public static func scan(
        _ candidates: [(id: Int32, url: URL, size: Int64)],
        progress: (@Sendable (DuplicateScanPhase, Int, Int) -> Void)? = nil
    ) async throws -> DuplicateScanResult {
        progress?(.grouping, 0, 0)
        var bySize: [Int64: [(id: Int32, url: URL)]] = [:]
        for candidate in candidates where candidate.size > 0 {
            try Task.checkCancellation()
            bySize[candidate.size, default: []].append((candidate.id, candidate.url))
        }

        let colliding = bySize.filter { $0.value.count > 1 }
        let totalBuckets = colliding.count
        progress?(.hashing, 0, totalBuckets)

        var groups: [DuplicateGroup] = []
        var fullContentHashCalls = 0
        var done = 0

        try await withThrowingTaskGroup(of: DuplicateScanResult.self) { taskGroup in
            for (size, files) in colliding {
                taskGroup.addTask {
                    try Task.checkCancellation()
                    return try hashAndGroup(files: files, size: size)
                }
            }
            for try await partial in taskGroup {
                try Task.checkCancellation()
                groups.append(contentsOf: partial.groups)
                fullContentHashCalls += partial.fullContentHashCalls
                done += 1
                progress?(.hashing, done, totalBuckets)
            }
        }

        progress?(.complete, totalBuckets, totalBuckets)
        return DuplicateScanResult(groups: groups, fullContentHashCalls: fullContentHashCalls)
    }

    private static func hashAndGroup(
        files: [(id: Int32, url: URL)],
        size: Int64
    ) throws -> DuplicateScanResult {
        var byPartialHash: [String: [(id: Int32, url: URL)]] = [:]
        for file in files {
            try Task.checkCancellation()
            guard let partial = try partialHash(url: file.url, bytes: 65_536) else { continue }
            byPartialHash[partial, default: []].append(file)
        }

        var groups: [DuplicateGroup] = []
        var fullContentHashCalls = 0
        for (_, collision) in byPartialHash where collision.count > 1 {
            let partitioned = partitionClones(collision)
            for cluster in partitioned.clusters {
                groups.append(DuplicateGroup(
                    hash: "shared-extents",
                    fileIDs: cluster.map(\.id).sorted(),
                    sizeEach: size,
                    sharesStorage: true
                ))
            }
            var byFullHash: [String: [Int32]] = [:]
            for file in partitioned.needsFullHash {
                try Task.checkCancellation()
                fullContentHashCalls += 1
                guard let full = try fullHash(url: file.url) else { continue }
                byFullHash[full, default: []].append(file.id)
            }
            for (hash, ids) in byFullHash where ids.count > 1 {
                groups.append(DuplicateGroup(
                    hash: hash,
                    fileIDs: ids.sorted(),
                    sizeEach: size,
                    sharesStorage: false
                ))
            }
        }
        return DuplicateScanResult(groups: groups, fullContentHashCalls: fullContentHashCalls)
    }

    /// Full extent-map matches become clone groups and never reach
    /// `fullHash`. Everyone else, including a clone that has been
    /// written since, is hashed.
    private static func partitionClones(
        _ files: [(id: Int32, url: URL)]
    ) -> (clusters: [[(id: Int32, url: URL)]], needsFullHash: [(id: Int32, url: URL)]) {
        var remaining = files
        var clusters: [[(id: Int32, url: URL)]] = []
        var needsFullHash: [(id: Int32, url: URL)] = []
        while let seed = remaining.first {
            remaining.removeFirst()
            var cluster = [seed]
            var rest: [(id: Int32, url: URL)] = []
            for other in remaining {
                if CloneDetector.areLikelyClones(seed.url.path, other.url.path) {
                    cluster.append(other)
                } else {
                    rest.append(other)
                }
            }
            remaining = rest
            if cluster.count > 1 {
                clusters.append(cluster)
            } else {
                needsFullHash.append(seed)
            }
        }
        return (clusters, needsFullHash)
    }

    /// First-pass filter only. MD5 is fine here because colliding files still
    /// go through streaming SHA256 before they become a duplicate group.
    private static func partialHash(url: URL, bytes: Int) throws -> String? {
        try Task.checkCancellation()
        guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
        defer { try? handle.close() }
        guard let data = try? handle.read(upToCount: bytes) else { return nil }
        try Task.checkCancellation()
        return Insecure.MD5.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    /// Streaming SHA256 — same digest as hashing a full `Data`, without
    /// holding the whole file in a contiguous buffer.
    private static func fullHash(url: URL) throws -> String? {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
        defer { try? handle.close() }
        var hasher = SHA256()
        let chunkSize = 1024 * 1024
        while true {
            try Task.checkCancellation()
            let chunk: Data?
            do {
                chunk = try handle.read(upToCount: chunkSize)
            } catch {
                return nil
            }
            // `read(upToCount:)` returns nil at EOF — that is success, not failure.
            guard let chunk, !chunk.isEmpty else { break }
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }
}

File: windows/src/DiskMap.Core/DuplicateFinder.cs
using System.Security.Cryptography;

namespace DiskMap.Core;

public sealed record DuplicateGroup(
    string Hash,
    List<int> FileIDs,
    long SizeEach,
    bool SharesStorage)
{
    /// <summary>
    /// Bytes a later confirm would actually free. Content copies each
    /// occupy their own blocks. Shared extents count once, and only when
    /// every copy in the group is being deleted.
    /// </summary>
    public long ReclaimableBytes(IReadOnlySet<int> deleting)
    {
        var removing = FileIDs.Where(deleting.Contains).ToList();
        if (removing.Count == 0 || SizeEach <= 0) return 0;
        if (SharesStorage)
            return removing.Count == FileIDs.Count ? SizeEach : 0;
        return SizeEach * removing.Count;
    }

    /// <summary>Oldest modified day, then lowest id — the file that stays unchecked.</summary>
    public int? DefaultKeeperId(Func<int, int> modifiedDay) =>
        FileIDs.Count == 0 ? null
            : FileIDs.MinBy(id => (modifiedDay(id), id));
}

internal sealed record DuplicateScanResult(List<DuplicateGroup> Groups, int FullContentHashCalls);

/// <summary>
/// Three-phase duplicate detection, cheapest checks first:
///
///   1. Group by exact logical size.
///   2. Within a size group, hash only the first 64 KB.
///   3. Only for files that still collide, hash full content —
///      unless CloneDetector.AreLikelyClones says they share a full
///      extent map. A clone overwritten past the 64 KB window still
///      collides on the partial hash and must be hashed.
/// </summary>
public static class DuplicateFinder
{
    public static async Task<List<DuplicateGroup>> FindDuplicatesAsync(
        List<(int Id, string Path, long Size)> candidates)
    {
        var result = await ScanAsync(candidates);
        return result.Groups;
    }

    /// <summary>
    /// Regular files under root, skipping directories, empty files, and
    /// cloud placeholders (opening those would recall the content).
    /// </summary>
    public static List<(int Id, string Path, long Size)> Candidates(FileTree tree, string root)
    {
        var result = new List<(int, string, long)>();
        var stack = new Stack<int>();
        if (tree.Count > 0) stack.Push(0);
        while (stack.Count > 0)
        {
            int id = stack.Pop();
            if (id > 0)
            {
                bool notDownloaded = (tree.Flags[id] & NodeFlags.NotDownloaded) != 0;
                if (!tree.IsDirectory[id] && !notDownloaded && tree.LogicalSize[id] > 0)
                    result.Add((id, tree.PathOf(id, root), tree.LogicalSize[id]));
            }
            int child = tree.FirstChild[id];
            while (child != -1)
            {
                stack.Push(child);
                child = tree.NextSibling[child];
            }
        }
        return result;
    }

    internal static async Task<DuplicateScanResult> ScanAsync(
        List<(int Id, string Path, long Size)> candidates)
    {
        var bySize = new Dictionary<long, List<(int Id, string Path)>>();
        foreach (var c in candidates)
        {
            if (c.Size <= 0) continue;
            if (!bySize.TryGetValue(c.Size, out var list))
                bySize[c.Size] = list = [];
            list.Add((c.Id, c.Path));
        }

        var groups = new List<DuplicateGroup>();
        int fullContentHashCalls = 0;
        var gate = new object();

        var tasks = bySize
            .Where(kv => kv.Value.Count > 1)
            .Select(kv => Task.Run(() => HashAndGroup(kv.Value, kv.Key)))
            .ToArray();
        foreach (var partial in await Task.WhenAll(tasks))
        {
            groups.AddRange(partial.Groups);
            fullContentHashCalls += partial.FullContentHashCalls;
        }
        return new DuplicateScanResult(groups, fullContentHashCalls);
    }

    private static DuplicateScanResult HashAndGroup(List<(int Id, string Path)> files, long size)
    {
        var byPartial = new Dictionary<string, List<(int Id, string Path)>>();
        foreach (var file in files)
        {
            if (PartialHash(file.Path, 65_536) is { } partial)
            {
                if (!byPartial.TryGetValue(partial, out var list))
                    byPartial[partial] = list = [];
                list.Add(file);
            }
        }

        var groups = new List<DuplicateGroup>();
        int fullContentHashCalls = 0;
        foreach (var (_, collision) in byPartial)
        {
            if (collision.Count < 2) continue;
            var (clusters, needsFullHash) = PartitionClones(collision);
            foreach (var cluster in clusters)
            {
                groups.Add(new DuplicateGroup(
                    Hash: "shared-extents",
                    FileIDs: cluster.Select(c => c.Id).OrderBy(x => x).ToList(),
                    SizeEach: size,
                    SharesStorage: true));
            }
            var byFull = new Dictionary<string, List<int>>();
            foreach (var file in needsFullHash)
            {
                fullContentHashCalls++;
                if (FullHash(file.Path) is { } full)
                {
                    if (!byFull.TryGetValue(full, out var list))
                        byFull[full] = list = [];
                    list.Add(file.Id);
                }
            }
            foreach (var (hash, ids) in byFull)
            {
                if (ids.Count > 1)
                {
                    groups.Add(new DuplicateGroup(
                        Hash: hash,
                        FileIDs: ids.OrderBy(x => x).ToList(),
                        SizeEach: size,
                        SharesStorage: false));
                }
            }
        }
        return new DuplicateScanResult(groups, fullContentHashCalls);
    }

    /// <summary>
    /// Full extent-map matches become clone groups and never reach
    /// FullHash. Everyone else, including a clone that has been written
    /// since, is hashed.
    /// </summary>
    private static (List<List<(int Id, string Path)>> Clusters, List<(int Id, string Path)> NeedsFullHash)
        PartitionClones(List<(int Id, string Path)> files)
    {
        var remaining = new List<(int Id, string Path)>(files);
        var clusters = new List<List<(int, string)>>();
        var needsFullHash = new List<(int, string)>();
        while (remaining.Count > 0)
        {
            var seed = remaining[0];
            remaining.RemoveAt(0);
            var cluster = new List<(int, string)> { seed };
            var rest = new List<(int, string)>();
            foreach (var other in remaining)
            {
                if (CloneDetector.AreLikelyClones(seed.Path, other.Path))
                    cluster.Add(other);
                else
                    rest.Add(other);
            }
            remaining = rest;
            if (cluster.Count > 1) clusters.Add(cluster);
            else needsFullHash.Add(seed);
        }
        return (clusters.Select(c => c.ToList()).ToList(), needsFullHash);
    }

    private static string? PartialHash(string path, int bytes)
    {
        try
        {
            using var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            var buffer = new byte[bytes];
            int read = stream.Read(buffer, 0, bytes);
            return Convert.ToHexString(SHA256.HashData(buffer.AsSpan(0, read)));
        }
        catch { return null; }
    }

    private static string? FullHash(string path)
    {
        try
        {
            using var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            return Convert.ToHexString(SHA256.HashData(stream));
        }
        catch { return null; }
    }
}
